File: RudderUnityPluginDemo/Assets/Plugins/iOS/RudderSDKUnity/RudderClientWrapper.m
//
//  RudderClientWrapper.m
//  Pods-RudderSDKUnityTest
//
//  Created by Arnab Pal on 15/12/19.
//

#import "RudderClientWrapper.h"
#import "RudderClient.h"
#import "Utils.h"
#import "RudderLogger.h"
#import "RudderElementCache.h"

static RudderClient *_rudderClient;

@implementation RudderClientWrapper

+ (void)_initiateInstance:(NSString *)_anonymousId
                 writeKey:(NSString *)_writeKey
              endPointUrl:(NSString *)_endPointUrl
           flushQueueSize:(int)_flushQueueSize
         dbCountThreshold:(int)_dbCountThreshold
             sleepTimeOut:(int)_sleepTimeout
                 logLevel:(int)_logLevel {
    if (_rudderClient == nil) {
        [RudderElementCache setAnonymousId:_anonymousId];
        RudderConfigBuilder *builder = [[RudderConfigBuilder alloc] init];
        [builder withEndPointUrl:_endPointUrl];
        [builder withFlushQueueSize:_flushQueueSize];
        [builder withDBCountThreshold:_dbCountThreshold];
        [builder withSleepTimeOut:_sleepTimeout];
        [builder withLoglevel:_logLevel];
        _rudderClient = [RudderClient getInstance:_writeKey config:[builder build]];
    }
}

+ (void)_logEvent:(NSString *)_eventType
        eventName:(NSString *)_eventName
   eventPropsJson:(NSString *)_eventPropsJson
    userPropsJson:(NSString *)_userPropsJson
      optionsJson:(NSString *)_optionsJson
{
    if (_rudderClient == nil) return;
    
    RudderMessageBuilder *builder = [[RudderMessageBuilder alloc] init];
    [builder setEventName:_eventName];
    [builder setPropertyDict:[self _convertToDict:_eventPropsJson]];
    [builder setUserProperty:[self _convertToDict:_userPropsJson]];
//    [builder setRudderOption:[self _convertToDict:_optionsJson]];
    
    if ([_eventType isEqualToString:@"track"]) {
        [_rudderClient trackMessage:[builder build]];
    } else if ([_eventType isEqualToString:@"screen"]) {
        [_rudderClient screenWithMessage:[builder build]];
    } else {
        [RudderLogger logError:@"message type is not supported"];
    }
    
}

+ (void)_identify:(NSString *)_userId traitsJson:(NSString *)_traitsJson optionsJson:(NSString *)_optionsJson {
    NSDictionary *traitsDict = [self _convertToDict:_traitsJson];
    if (traitsDict == nil) {
        // if traits is not filled in, fill with anonymousId
        traitsDict = @{@"anonymousId": [[[[UIDevice currentDevice] identifierForVendor] UUIDString]lowercaseString]};
    } else {
        // if anonymousId is not filled in
        NSString *anonymoysId = [traitsDict valueForKey:@"anonymousId"];
        if (anonymoysId == nil) {
            [[traitsDict mutableCopy] setObject:[[[[UIDevice currentDevice] identifierForVendor] UUIDString]lowercaseString] forKey:@"anonymousId"];
        }
    }
    NSDictionary *optinsDict = [self _convertToDict:_optionsJson];
    // make the identify call
    [_rudderClient identify:_userId traits:traitsDict options:optinsDict];
}

+ (void)_reset {
    if (_rudderClient == nil) return;
    [_rudderClient reset];
}

+ (NSDictionary<NSString*, NSObject*>*) _convertToDict: (NSString*) _json {
    NSStringEncoding encoding = 0;
    NSData *jsonData = [_json dataUsingEncoding:encoding];
    NSError *error = nil;
    NSDictionary *parsedData = [NSJSONSerialization JSONObjectWithData:jsonData options:kNilOptions error:&error];
    if (error == nil) {
        return parsedData;
    } else {
        return nil;
    }
}

@end

File: Plugins/Android/rudder-sdk-unity-plugin-android/src/main/java/com/rudderlabs/android/sdk/core/RudderClientWrapper.java
package com.rudderlabs.android.sdk.core;

import android.content.Context;
import android.text.TextUtils;

import com.rudderlabs.android.sdk.core.util.Utils;

import java.util.Locale;
import java.util.Map;

public class RudderClientWrapper {

    private static RudderClient rudderClient;

    public static void _initiateInstance(
            Context _context,
            String _anonymousId,
            String _writeKey,
            String _endPointUrl,
            int _flushQueueSize,
            int _dbCountThreshold,
            int _sleepTimeout,
            int _logLevel
    ) {
        if (_context == null) {
            RudderLogger.logError("Context can not be null");
            return;
        }

        if (TextUtils.isEmpty(_writeKey)) {
            RudderLogger.logError("WriteKey can not be null or empty");
            return;
        }

        if (TextUtils.isEmpty(_anonymousId)) {
            _anonymousId = Utils.getDeviceId(_context);
        }
        RudderElementCache.setAnonymousId(_anonymousId);

        RudderConfig config = new RudderConfig.Builder()
                .withEndPointUri(_endPointUrl)
                .withFlushQueueSize(_flushQueueSize)
                .withDbThresholdCount(_dbCountThreshold)
                .withSleepCount(_sleepTimeout)
                .withLogLevel(_logLevel)
                .build();

        rudderClient = RudderClient.getInstance(_context, _writeKey, config);
        RudderLogger.logDebug("Client initiated successfully");
    }

    public static void _logEvent(
            String _eventType,
            String _eventName,
            String _eventPropsJson,
            String _userPropsJson,
            String _optionsJson
    ) {
        if (rudderClient == null) {
            return;
        }

        RudderMessage message = new RudderMessageBuilder()
                .setEventName(_eventName)
                .setProperty(Utils.convertToMap(_eventPropsJson))
                .setUserProperty(Utils.convertToMap(_userPropsJson))
                .setRudderOption(Utils.convertToMap(_optionsJson))
                .build();

        switch (_eventType) {
            case "track":
                rudderClient.track(message);
                break;
            case "screen":
                rudderClient.screen(message);
                break;
            case "identify":
                RudderLogger.logError("Identify is not supported via this endpoint");
        }
    }

    public static void _identify(String _userId, String _traitsJson, String _optionsJson) {
        RudderLogger.logDebug(String.format(Locale.US, "_userId: %s", _userId));
        RudderLogger.logDebug(String.format(Locale.US, "_traitsJson: %s", _traitsJson));
        RudderLogger.logDebug(String.format(Locale.US, "_optionsJson: %s", _optionsJson));

        if (rudderClient == null) {
            return;
        }

        // check for anonymousId
        Map<String, Object> traitsMap = Utils.convertToMap(_traitsJson);
        if (!traitsMap.containsKey("anonymousId")) {
            traitsMap.put("anonymousId", Utils.getDeviceId(rudderClient.getApplication()));
        }
        rudderClient.identify(
                _userId,
                traitsMap,
                Utils.convertToMap(_optionsJson)
        );
    }

    public static void _reset() {
        if (rudderClient == null) {
            return;
        }

        rudderClient.reset();
    }
}
